// @einlogic/mcp-fabric-api/build/auth/token-manager.js
import { DefaultAzureCredential, DeviceCodeCredential, ClientSecretCredential, InteractiveBrowserCredential, } from "@azure/identity";
import { DeviceCodeAuthRequired } from "../core/errors.js";
const FABRIC_SCOPE = "https://api.fabric.microsoft.com/.default";
const POWERBI_SCOPE = "https://analysis.windows.net/powerbi/api/.default";
const DATABASE_SCOPE = "https://database.windows.net/.default";
const KUSTO_SCOPE = "https://api.kusto.windows.net/.default";
const REFRESH_BUFFER_MS = 5 * 60 * 1000; // 5 minutes
export class TokenManager {
    credential;
    cache = new Map();
    currentTenantId;
    authConfig;
    /** Background device-code auth promises keyed by scope URL */
    pendingDeviceCodeAuth = new Map();
    /** Latest device code message for display to user */
    lastDeviceCodeMessage = null;
    constructor(config) {
        this.authConfig = config ?? { method: "default" };
        this.currentTenantId = this.authConfig.tenantId;
        this.credential = this.createCredential(this.currentTenantId);
    }
    createCredential(tenantId) {
        const { method, clientId, clientSecret } = this.authConfig;
        switch (method) {
            case "device-code": {
                if (!clientId || !tenantId) {
                    throw new Error("AUTH_METHOD=device-code requires AZURE_CLIENT_ID and AZURE_TENANT_ID");
                }
                const { onDeviceCodePrompt } = this.authConfig;
                return new DeviceCodeCredential({
                    tenantId,
                    clientId,
                    userPromptCallback: (info) => {
                        console.error(`\n[Auth] ${info.message}\n`);
                        this.lastDeviceCodeMessage = info.message;
                        if (onDeviceCodePrompt) {
                            onDeviceCodePrompt(info.message);
                        }
                    },
                });
            }
            case "client-secret": {
                if (!tenantId || !clientId || !clientSecret) {
                    throw new Error("AUTH_METHOD=client-secret requires AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET");
                }
                return new ClientSecretCredential(tenantId, clientId, clientSecret);
            }
            case "interactive-browser": {
                return new InteractiveBrowserCredential({
                    tenantId,
                    clientId,
                });
            }
            default: {
                return tenantId
                    ? new DefaultAzureCredential({ tenantId })
                    : new DefaultAzureCredential();
            }
        }
    }
    switchTenant(tenantId) {
        this.currentTenantId = tenantId;
        this.credential = this.createCredential(tenantId);
        this.cache.clear();
        this.pendingDeviceCodeAuth.clear();
        this.lastDeviceCodeMessage = null;
    }
    clearCache() {
        this.cache.clear();
    }
    getCurrentTenantId() {
        return this.currentTenantId;
    }
    getCachedToken(scope) {
        const scopeUrl = scope === "fabric" ? FABRIC_SCOPE : scope === "powerbi" ? POWERBI_SCOPE : scope === "kusto" ? KUSTO_SCOPE : DATABASE_SCOPE;
        return this.cache.get(scopeUrl);
    }
    async getToken(scope) {
        const scopeUrl = scope === "fabric" ? FABRIC_SCOPE : scope === "powerbi" ? POWERBI_SCOPE : scope === "kusto" ? KUSTO_SCOPE : DATABASE_SCOPE;
        const cached = this.cache.get(scopeUrl);
        if (cached && cached.expiresOnTimestamp - Date.now() > REFRESH_BUFFER_MS) {
            return cached.token;
        }
        // For device-code flow, start auth in the background and return the code to the user immediately
        if (this.authConfig.method === "device-code") {
            return this.getTokenWithDeviceCode(scopeUrl);
        }
        try {
            const token = await this.credential.getToken(scopeUrl);
            if (!token) {
                throw new Error(`Failed to acquire token for scope: ${scopeUrl}`);
            }
            this.cache.set(scopeUrl, token);
            return token.token;
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            const method = this.authConfig.method;
            if (method === "client-secret") {
                throw new Error(`Client secret authentication failed. Verify AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET are correct.\nDetails: ${message}`);
            }
            if (method === "interactive-browser") {
                throw new Error(`Interactive browser authentication failed. Ensure a browser is accessible.\nDetails: ${message}`);
            }
            if (message.includes("az login") || message.includes("DefaultAzureCredential")) {
                throw new Error(`Azure authentication failed. Run 'az login' first to authenticate.\nDetails: ${message}`);
            }
            throw error;
        }
    }
    /**
     * Device-code specific token acquisition.
     * Starts polling in the background and throws DeviceCodeAuthRequired immediately
     * so the tool can return the auth URL/code to the user instead of blocking.
     */
    async getTokenWithDeviceCode(scopeUrl) {
        // If there's already a pending auth for this scope, check if it completed
        const pending = this.pendingDeviceCodeAuth.get(scopeUrl);
        if (pending) {
            // Race: either the background auth completed, or we time out quickly
            const POLL_TIMEOUT_MS = 3000;
            const result = await Promise.race([
                pending.then((token) => ({ resolved: true, token })),
                new Promise((resolve) => setTimeout(() => resolve({ resolved: false }), POLL_TIMEOUT_MS)),
            ]);
            if (result.resolved) {
                this.cache.set(scopeUrl, result.token);
                this.pendingDeviceCodeAuth.delete(scopeUrl);
                this.lastDeviceCodeMessage = null;
                return result.token.token;
            }
            // Still waiting — remind the user
            throw new DeviceCodeAuthRequired(this.lastDeviceCodeMessage ?? "Device code authentication is in progress. Please complete sign-in in your browser.");
        }
        // No pending auth — start a new one in the background
        this.lastDeviceCodeMessage = null;
        const authPromise = this.credential.getToken(scopeUrl).then((token) => {
            if (!token)
                throw new Error(`Failed to acquire token for scope: ${scopeUrl}`);
            return token;
        });
        this.pendingDeviceCodeAuth.set(scopeUrl, authPromise);
        // Handle background completion/failure
        authPromise
            .then((token) => {
            this.cache.set(scopeUrl, token);
            this.pendingDeviceCodeAuth.delete(scopeUrl);
            this.lastDeviceCodeMessage = null;
            console.error(`[Auth] Device code authentication completed for scope: ${scopeUrl}`);
        })
            .catch((err) => {
            this.pendingDeviceCodeAuth.delete(scopeUrl);
            console.error(`[Auth] Device code authentication failed: ${err instanceof Error ? err.message : String(err)}`);
        });
        // Wait briefly for the device code callback to fire so we can capture the message
        const CALLBACK_WAIT_MS = 3000;
        await new Promise((resolve) => setTimeout(resolve, CALLBACK_WAIT_MS));
        // If the auth already completed (e.g., cached token in the credential), return it
        const nowCached = this.cache.get(scopeUrl);
        if (nowCached && nowCached.expiresOnTimestamp - Date.now() > REFRESH_BUFFER_MS) {
            return nowCached.token;
        }
        // Throw the device code info back to the caller so it surfaces in the tool response
        throw new DeviceCodeAuthRequired(this.lastDeviceCodeMessage ?? "Device code authentication started. Please check the MCP server logs for the sign-in URL and code.");
    }
    async getFabricToken() {
        return this.getToken("fabric");
    }
    async getPowerBIToken() {
        return this.getToken("powerbi");
    }
    async getDatabaseToken() {
        return this.getToken("database");
    }
    async getKustoToken() {
        return this.getToken("kusto");
    }
}
