// @einlogic/mcp-fabric-api/build/utils/file-utils.js
import { readFile, writeFile, readdir, mkdir, stat } from "fs/promises";
import { resolve, relative, extname, dirname, join } from "path";
export async function readContentFromFile(filePath) {
    const absolutePath = resolve(filePath);
    const fileStat = await stat(absolutePath);
    if (!fileStat.isFile()) {
        throw new Error(`Path is not a file: ${absolutePath}`);
    }
    return readFile(absolutePath, "utf-8");
}
export async function readFilesFromDirectory(dirPath, extensions) {
    const absolutePath = resolve(dirPath);
    const dirStat = await stat(absolutePath);
    if (!dirStat.isDirectory()) {
        throw new Error(`Path is not a directory: ${absolutePath}`);
    }
    const entries = [];
    const lowerExtensions = extensions?.map((e) => e.toLowerCase());
    async function walk(dir) {
        const items = await readdir(dir, { withFileTypes: true });
        for (const item of items) {
            const fullPath = resolve(dir, item.name);
            if (item.isDirectory()) {
                await walk(fullPath);
            }
            else if (item.isFile()) {
                if (lowerExtensions && !lowerExtensions.includes(extname(item.name).toLowerCase())) {
                    continue;
                }
                const relPath = relative(absolutePath, fullPath).replace(/\\/g, "/");
                const content = await readFile(fullPath, "utf-8");
                entries.push({ path: relPath, content });
            }
        }
    }
    await walk(absolutePath);
    if (entries.length === 0) {
        const filterMsg = extensions ? ` matching extensions: ${extensions.join(", ")}` : "";
        throw new Error(`No files found in directory: ${absolutePath}${filterMsg}`);
    }
    return entries;
}
export async function writeContentToFile(filePath, content) {
    const absolutePath = resolve(filePath);
    await mkdir(dirname(absolutePath), { recursive: true });
    await writeFile(absolutePath, content, "utf-8");
    return absolutePath;
}
export async function writeFilesToDirectory(dirPath, files) {
    const absolutePath = resolve(dirPath);
    const written = [];
    for (const file of files) {
        const fullPath = join(absolutePath, file.path);
        await mkdir(dirname(fullPath), { recursive: true });
        await writeFile(fullPath, file.content, "utf-8");
        written.push(file.path);
    }
    return written;
}
export async function resolveContentOrFile(content, filePath, fieldName) {
    if (content && filePath) {
        throw new Error(`Provide either '${fieldName}' or '${fieldName}FilePath', not both`);
    }
    if (filePath) {
        return readContentFromFile(filePath);
    }
    if (content) {
        return content;
    }
    throw new Error(`Either '${fieldName}' or '${fieldName}FilePath' must be provided`);
}
export async function resolveFilesOrDirectory(files, dirPath, extensions) {
    if (files && files.length > 0 && dirPath) {
        throw new Error("Provide either 'files' or a directory path, not both");
    }
    if (dirPath) {
        return readFilesFromDirectory(dirPath, extensions);
    }
    if (files && files.length > 0) {
        return files;
    }
    throw new Error("Either 'files' or a directory path must be provided");
}
