// @einlogic/mcp-fabric-api/build/utils/logger.js
const currentLevel = process.env.LOG_LEVEL?.toLowerCase() === "debug" ? "debug" : "info";
function timestamp() {
    return new Date().toISOString();
}
function format(level, component, message, extra) {
    const parts = [`[${timestamp()}] [${level.toUpperCase()}] [${component}] ${message}`];
    if (extra && Object.keys(extra).length > 0) {
        parts.push(JSON.stringify(extra));
    }
    return parts.join(" ");
}
export const logger = {
    info(component, message, extra) {
        console.error(format("info", component, message, extra));
    },
    debug(component, message, extra) {
        if (currentLevel === "debug") {
            console.error(format("debug", component, message, extra));
        }
    },
    warn(component, message, extra) {
        console.error(format("warn", component, message, extra));
    },
    error(component, message, extra) {
        console.error(format("error", component, message, extra));
    },
    isDebug() {
        return currentLevel === "debug";
    },
};
/**
 * Redact sensitive headers (Authorization) for safe logging.
 * Returns a plain object with header names lowercased.
 */
export function safeHeaders(headers) {
    const result = {};
    headers.forEach((value, key) => {
        const lower = key.toLowerCase();
        if (lower === "authorization") {
            result[lower] = "Bearer [REDACTED]";
        }
        else {
            result[lower] = value;
        }
    });
    return result;
}
/**
 * Summarize definition parts for debug logging without exposing payload content.
 * Logs part path, payloadType, and payload byte size only.
 */
export function summarizeDefinitionParts(parts) {
    return parts.map((p) => ({
        path: p.path,
        payloadType: p.payloadType,
        payloadBytes: p.payload.length,
    }));
}
