// @einlogic/mcp-fabric-api/build/tools/sql-endpoint.js
import { z } from "zod";
import { formatToolError } from "../core/errors.js";
import { paginateAll } from "../core/pagination.js";
const READ = { readOnlyHint: true, destructiveHint: false };
const WRITE = { readOnlyHint: false, destructiveHint: false };
const DESTRUCTIVE = { readOnlyHint: false, destructiveHint: true };
export function registerSqlEndpointTools(server, fabricClient, sqlClient) {
    server.tool("sql_endpoint_list", "List all SQL endpoints in a workspace", { workspaceId: z.string().describe("The workspace ID") }, READ, async ({ workspaceId }) => {
        try {
            const endpoints = await paginateAll(fabricClient, `/workspaces/${workspaceId}/sqlEndpoints`);
            return { content: [{ type: "text", text: JSON.stringify(endpoints, null, 2) }] };
        }
        catch (error) {
            return formatToolError(error);
        }
    });
    server.tool("sql_endpoint_get", "Get details of a specific SQL endpoint, including connection string", {
        workspaceId: z.string().describe("The workspace ID"),
        sqlEndpointId: z.string().describe("The SQL endpoint ID"),
    }, READ, async ({ workspaceId, sqlEndpointId }) => {
        try {
            const response = await fabricClient.get(`/workspaces/${workspaceId}/sqlEndpoints/${sqlEndpointId}`);
            return { content: [{ type: "text", text: JSON.stringify(response.data, null, 2) }] };
        }
        catch (error) {
            return formatToolError(error);
        }
    });
    server.tool("sql_endpoint_get_connection_string", "Get the TDS connection string for a SQL endpoint", {
        workspaceId: z.string().describe("The workspace ID"),
        sqlEndpointId: z.string().describe("The SQL endpoint ID"),
    }, READ, async ({ workspaceId, sqlEndpointId }) => {
        try {
            const response = await fabricClient.get(`/workspaces/${workspaceId}/sqlEndpoints/${sqlEndpointId}`);
            const properties = response.data.properties;
            const connectionString = properties?.connectionString;
            if (!connectionString) {
                return { content: [{ type: "text", text: "No connection string available for this SQL endpoint. The endpoint may still be provisioning." }] };
            }
            return { content: [{ type: "text", text: JSON.stringify({ connectionString }, null, 2) }] };
        }
        catch (error) {
            return formatToolError(error);
        }
    });
    server.tool("sql_endpoint_execute_query", "Execute a T-SQL query against a lakehouse or warehouse SQL endpoint", {
        query: z.string().describe("T-SQL query to execute"),
        workspaceId: z.string().describe("The workspace ID containing the resource"),
        itemId: z.string().describe("The lakehouse or warehouse ID"),
        itemType: z.enum(["Lakehouse", "Warehouse"]).describe("Type of the item to query"),
        database: z.string().optional().describe("Database name (defaults to the item's display name)"),
        maxRows: z.number().optional().describe("Maximum rows to return (default: 1000)"),
    }, WRITE, async ({ query, workspaceId, itemId, itemType, database, maxRows }) => {
        try {
            let connectionString;
            let displayName;
            if (itemType === "Lakehouse") {
                const response = await fabricClient.get(`/workspaces/${workspaceId}/lakehouses/${itemId}`);
                displayName = response.data.displayName;
                const properties = response.data.properties;
                const sqlEndpointProps = properties?.sqlEndpointProperties;
                connectionString = sqlEndpointProps?.connectionString;
            }
            else {
                const response = await fabricClient.get(`/workspaces/${workspaceId}/warehouses/${itemId}`);
                displayName = response.data.displayName;
                const properties = response.data.properties;
                connectionString = properties?.connectionString;
            }
            if (!connectionString) {
                return { content: [{ type: "text", text: `No SQL endpoint connection string available for this ${itemType.toLowerCase()}. It may still be provisioning.` }] };
            }
            const server = connectionString.replace(/^.*?:\/\//, "").replace(/,.*$/, "").replace(/;.*$/, "");
            const dbName = database ?? displayName ?? itemId;
            const result = await sqlClient.executeQuery(server, dbName, query, maxRows);
            return { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] };
        }
        catch (error) {
            return formatToolError(error);
        }
    });
}
