// @einlogic/mcp-fabric-api/build/core/errors.js
/**
 * Thrown when device-code authentication is required.
 * The message contains the URL and code the user must use to authenticate.
 * This is NOT an error — it signals the user needs to act before retrying.
 */
export class DeviceCodeAuthRequired extends Error {
    deviceCodeMessage;
    constructor(deviceCodeMessage) {
        super(deviceCodeMessage);
        this.deviceCodeMessage = deviceCodeMessage;
        this.name = "DeviceCodeAuthRequired";
    }
}
export class FabricApiError extends Error {
    statusCode;
    errorCode;
    relatedResource;
    requestId;
    errorDetails;
    constructor(message, statusCode, errorCode, relatedResource, requestId, errorDetails) {
        super(message);
        this.statusCode = statusCode;
        this.errorCode = errorCode;
        this.relatedResource = relatedResource;
        this.requestId = requestId;
        this.errorDetails = errorDetails;
        this.name = "FabricApiError";
    }
}
export function formatToolError(error) {
    if (error instanceof DeviceCodeAuthRequired) {
        return {
            content: [{ type: "text", text: `🔐 Authentication required.\n\n${error.deviceCodeMessage}\n\nOnce you have completed the sign-in, retry the tool call.` }],
            isError: false,
        };
    }
    let message;
    if (error instanceof FabricApiError) {
        message = `Fabric API Error (${error.statusCode}): ${error.message}`;
        if (error.errorCode)
            message += `\nError code: ${error.errorCode}`;
        if (error.relatedResource)
            message += `\nRelated resource: ${error.relatedResource}`;
        if (error.requestId)
            message += `\nRequest ID: ${error.requestId}`;
        if (error.errorDetails && error.errorDetails.length > 0) {
            message += `\nDetails: ${JSON.stringify(error.errorDetails)}`;
        }
    }
    else if (error instanceof Error) {
        message = error.message;
    }
    else {
        message = String(error);
    }
    return {
        content: [{ type: "text", text: message }],
        isError: true,
    };
}
