// @einlogic/mcp-fabric-api/build/core/job-scheduler.js
export async function runOnDemandJob(client, workspaceId, itemId, jobType, executionData) {
    const body = {};
    if (executionData) {
        body.executionData = executionData;
    }
    const response = await client.post(`/workspaces/${workspaceId}/items/${itemId}/jobs/instances?jobType=${jobType}`, Object.keys(body).length > 0 ? body : undefined);
    // 202 Accepted with no body — extract job instance ID from Location header
    if (response.data === undefined || response.data === null) {
        const jobInstanceId = response.lro?.location?.match(/instances\/([^/?]+)/)?.[1];
        return {
            id: jobInstanceId ?? response.lro?.operationId ?? "unknown",
            itemId,
            jobType,
            invokeType: "OnDemand",
            status: "NotStarted",
        };
    }
    return response.data;
}
export async function getJobInstance(client, workspaceId, itemId, jobInstanceId) {
    const response = await client.get(`/workspaces/${workspaceId}/items/${itemId}/jobs/instances/${jobInstanceId}`);
    return response.data;
}
export async function cancelJobInstance(client, workspaceId, itemId, jobInstanceId) {
    await client.post(`/workspaces/${workspaceId}/items/${itemId}/jobs/instances/${jobInstanceId}/cancel`);
}
export async function listJobInstances(client, workspaceId, itemId) {
    const response = await client.get(`/workspaces/${workspaceId}/items/${itemId}/jobs/instances`);
    return response.data.value ?? [];
}
