// @einlogic/mcp-fabric-api/build/client/kusto-client.js
import { logger } from "../utils/logger.js";
const DEFAULT_MAX_ROWS = 1000;
const COMPONENT = "KustoClient";
export class KustoClient {
    tokenManager;
    constructor(tokenManager) {
        this.tokenManager = tokenManager;
    }
    async executeQuery(queryServiceUri, database, kql, maxRows = DEFAULT_MAX_ROWS) {
        const token = await this.tokenManager.getKustoToken();
        const url = `${queryServiceUri.replace(/\/+$/, "")}/v2/rest/query`;
        const body = {
            db: database,
            csl: kql,
            properties: {
                Options: {
                    servertimeout: "00:00:30",
                    truncationmaxrecords: maxRows,
                },
            },
        };
        logger.debug(COMPONENT, `POST ${url}`, { database, maxRows });
        const startTime = Date.now();
        const response = await fetch(url, {
            method: "POST",
            headers: {
                Authorization: `Bearer ${token}`,
                "Content-Type": "application/json; charset=utf-8",
                Accept: "application/json",
            },
            body: JSON.stringify(body),
        });
        const durationMs = Date.now() - startTime;
        if (!response.ok) {
            let errorDetail;
            try {
                const errorBody = await response.json();
                errorDetail = errorBody?.error
                    ? JSON.stringify(errorBody.error)
                    : JSON.stringify(errorBody);
            }
            catch {
                errorDetail = await response.text();
            }
            logger.error(COMPONENT, `KQL query failed`, {
                status: response.status,
                database,
                durationMs,
                errorDetail,
            });
            throw new Error(`KQL query failed (HTTP ${response.status}): ${errorDetail}`);
        }
        const frames = (await response.json());
        const primaryResult = frames.find((f) => f.FrameType === "DataTable" && f.TableKind === "PrimaryResult");
        if (!primaryResult) {
            throw new Error("KQL response did not contain a PrimaryResult table");
        }
        const columns = (primaryResult.Columns ?? []).map((c) => c.ColumnName);
        const rawRows = primaryResult.Rows ?? [];
        const cappedRows = rawRows.slice(0, maxRows);
        const rows = cappedRows.map((row) => {
            const record = {};
            for (let i = 0; i < columns.length; i++) {
                record[columns[i]] = row[i];
            }
            return record;
        });
        logger.debug(COMPONENT, `KQL query completed`, {
            status: response.status,
            database,
            durationMs,
            columnCount: columns.length,
            rowCount: rows.length,
        });
        return { columns, rows, rowCount: rows.length };
    }
}
