// @einlogic/mcp-fabric-api/build/client/sql-client.js
import { Connection, Request } from "tedious";
import { logger } from "../utils/logger.js";
const DEFAULT_MAX_ROWS = 1000;
const QUERY_TIMEOUT_MS = 30000;
const COMPONENT = "SqlClient";
export class SqlClient {
    tokenManager;
    constructor(tokenManager) {
        this.tokenManager = tokenManager;
    }
    async executeQuery(server, database, query, maxRows = DEFAULT_MAX_ROWS) {
        const token = await this.tokenManager.getDatabaseToken();
        logger.debug(COMPONENT, `Connecting to ${server}`, { database, maxRows });
        const startTime = Date.now();
        return new Promise((resolve, reject) => {
            const columns = [];
            const rows = [];
            const connection = new Connection({
                server,
                authentication: {
                    type: "azure-active-directory-access-token",
                    options: {
                        token,
                    },
                },
                options: {
                    database,
                    encrypt: true,
                    port: 1433,
                    requestTimeout: QUERY_TIMEOUT_MS,
                    trustServerCertificate: false,
                },
            });
            connection.on("connect", (err) => {
                if (err) {
                    const durationMs = Date.now() - startTime;
                    logger.error(COMPONENT, `Connection failed to ${server}`, { database, durationMs, error: err.message });
                    reject(new Error(`SQL connection failed: ${err.message}`));
                    return;
                }
                logger.debug(COMPONENT, `Connected to ${server}`, { database, durationMs: Date.now() - startTime });
                const request = new Request(query, (reqErr, rowCount) => {
                    const durationMs = Date.now() - startTime;
                    connection.close();
                    if (reqErr) {
                        logger.error(COMPONENT, `Query failed on ${server}`, { database, durationMs, error: reqErr.message });
                        reject(new Error(`SQL query failed: ${reqErr.message}`));
                        return;
                    }
                    logger.debug(COMPONENT, `Query completed on ${server}`, {
                        database,
                        durationMs,
                        columnCount: columns.length,
                        rowCount: rowCount ?? rows.length,
                    });
                    resolve({ columns, rows, rowCount: rowCount ?? rows.length });
                });
                request.on("columnMetadata", (columnMetadata) => {
                    const cols = Array.isArray(columnMetadata)
                        ? columnMetadata
                        : Object.values(columnMetadata);
                    for (const col of cols) {
                        columns.push(col.colName);
                    }
                });
                request.on("row", (rowColumns) => {
                    if (rows.length >= maxRows)
                        return;
                    const row = {};
                    const cols = Array.isArray(rowColumns)
                        ? rowColumns
                        : Object.values(rowColumns);
                    for (const col of cols) {
                        row[col.metadata.colName] = col.value;
                    }
                    rows.push(row);
                });
                connection.execSql(request);
            });
            connection.on("error", (err) => {
                logger.error(COMPONENT, `Connection error on ${server}`, { database, error: err.message });
                reject(new Error(`SQL connection error: ${err.message}`));
            });
            connection.connect();
        });
    }
}
